agent/php_user_instrument_wraprec_hashmap.c

```diff
@@ -0,0 +1,489 @@
+  nr_scope_bucket_t* bucket = (nr_scope_bucket_t*)nr_malloc(sizeof(nr_scope_bucket_t));
+  bucket->prev = NULL;
+  bucket->next = hashmap->buckets[hash_key];
+  bucket->key = (nr_scope_hashmap_key_t*)nr_malloc(sizeof(nr_scope_hashmap_key_t));
+  bucket->key->name = nr_strndup(key->name, key->name_len);
+  bucket->key->name_len = key->name_len;
+  bucket->key->name_hash = key->name_hash;
+  bucket->scoped_funcs_ht = nr_func_hashmap_create_internal(0);
+
+  if (hashmap->buckets[hash_key]) {
+    hashmap->buckets[hash_key]->prev = bucket;
+  }
+
+  hashmap->buckets[hash_key] = bucket;
+  ++hashmap->elements;
+
+  return bucket->scoped_funcs_ht;
+}
+
+static nr_func_hashmap_t* nr_scope_hashmap_lookup_internal(nr_scope_hashmap_t* hashmap, nr_scope_hashmap_key_t* key) {
+  size_t hash;
+  nr_scope_bucket_t* bucket = NULL;
+
+  if (nrunlikely((NULL == hashmap) || (NULL == key))) {
+    return NULL;
+  }
+
+  hash = nr_scope_hashmap_hash_key(hashmap->log2_num_buckets, key);
+  if (nr_scope_hashmap_fetch_internal(hashmap, hash, key, &bucket)) {
+    return bucket->scoped_funcs_ht;
+  }
+
+  return NULL;
+}
+
+static nr_func_hashmap_t* nr_scope_hashmap_update_internal(nr_scope_hashmap_t* hashmap, nr_scope_hashmap_key_t* key) {
+  size_t hash;
+  nr_scope_bucket_t* bucket = NULL;
+
+  if (nrunlikely((NULL == hashmap) || (NULL == key))) {
+    return NULL;
+  }
+
+  hash = nr_scope_hashmap_hash_key(hashmap->log2_num_buckets, key);
+  if (nr_scope_hashmap_fetch_internal(hashmap, hash, key, &bucket)) {
+    return bucket->scoped_funcs_ht;
+  }
+
+  return nr_scope_hashmap_add_internal(hashmap, hash, key);
+}
+
+static void nr_scope_hashmap_destroy_bucket_internal(nr_scope_bucket_t** bucket_ptr) {
+  nr_scope_bucket_t* bucket = *bucket_ptr;
+  if (NULL != bucket->key) {
+    nr_free(bucket->key->name);
+  }
+  nr_free(bucket->key);
+  nr_func_hashmap_destroy_internal(&bucket->scoped_funcs_ht);
+  nr_realfree((void**)bucket_ptr);
+}
+
+static void nr_scope_hashmap_destroy_internal(nr_scope_hashmap_t** hashmap_ptr) {
+  size_t count;
+  nr_scope_hashmap_t* hashmap = NULL;
+  size_t i;
+
+  if ((NULL == hashmap_ptr) || (NULL == *hashmap_ptr)) {
+    return;
+  }
+  hashmap = *hashmap_ptr;
+
+  count = (size_t)(1 << hashmap->log2_num_buckets);
+  for (i = 0; i < count; i++) {
+    nr_scope_bucket_t* bucket = hashmap->buckets[i];
+
+    while (bucket) {
+      nr_scope_bucket_t* next = bucket->next;
+
+      nr_scope_hashmap_destroy_bucket_internal(&bucket);
+
+      bucket = next;
+    }
+  }
+
+  nr_free(hashmap->buckets);
+  nr_realfree((void**)hashmap_ptr);
+}
+
+nr_func_hashmap_t* global_funcs_ht = NULL;
+nr_scope_hashmap_t* scope_ht = NULL;
+
+static void nr_php_user_instrument_wraprec_hashmap_name2keys(
+  nr_func_hashmap_key_t* func,
+  nr_scope_hashmap_key_t* scope,
+  const char* full_name,
+  int full_name_len) {
+
+  if (nrunlikely(NULL == full_name)) {
+    return;
+  }
+  if (nrunlikely(full_name_len <= 0)) {
+    return;
+  }
```

Comment:
Since this is only called from `nr_php_user_instrument_wraprec_hashmap_add` which already validates name and name_len, could we remove the extra checks and add the extra comment like `nr_php_user_instrument_wraprec_hashmap_add` has?

(here's the snippet since I can't link it directly because github keeps hiding the large diff)
```
/* This function expects namestr and namestrlen to be validated with
 * nr_php_user_instrument_is_name_valid() before being passed in.
 * Specifically, it requires that:
 * - namestrlen be greater than 0
 * - namestr must not be NULL and must not end with `:` (colon) . */

nruserfn_t* nr_php_user_instrument_wraprec_hashmap_add(const char* namestr, size_t namestrlen) {
```

Reply:
See c6c50d97dc5cf67311a4b504a7173d02223b8486.